README.md
atd-openoffice
==============

This is an English grammar checker for OpenOffice.org. Checking is performed by the After the Deadline proofreading software service.

Automattic no longer supports this extension.  We're putting it on Github so that you can feel free to fork it, hack it, and release your own version.

### Requirements:

* OpenOffice.org 3.2.0 or later
* Sun Java 1.5 or later.
* An internet connection

### Installation:

1. Go to Tools -> Extension Manager 
2. Cick Add... and select the atd-openoffice.oxt file
3. Click OK.
4. Restart OpenOffice.org (necessary to intialize the updated grammar checker)

### Use:

Visit Tools -> Extension Manager -> After the Deadline -> Options. Here you may modify 
the style checker settings and change the server this extension posts to. 

This extension checks your grammar (and style when enabled). It does not check your 
spelling. This functionality is left to the built-in spell checker.

There are two ways to check your grammar, as you type or via the grammar checker dialog:

Type your text and hit enter. Errors in the last paragraph will show a blue squiggly line.
Right click on an error and select an option to change it.

Select Tools -> Spelling and Grammar to bring up the grammar and spell checker dialog. Click
Ignore Rule to ignore all errors associated with the rule. Ignored rules last until you
restart OpenOffice.org. You may also select Tools -> Reset Ignored Rules to clear the ignored 
rules.

Settings are stored in ~/.AtD-OpenOffice.org. Delete this file if you want to clear your
AtD settings.

### Commercial use and running your own server

This extension requires a running instance of an [After the Deadline Server](https://github.com/automattic/atd-server) in order to work.  Automattic operates an that you can use for personal use as long as you don't send too much traffic.  The extension is configured to use this instance by default.

For high volume and commercial uses of AtD, you must run your own server.  The code is available on Github: [After the Deadline Server](https://github.com/automattic/atd-server).  See the [After the Deadline Developer's page](http://open.afterthedeadline.com/) for more information, and check out the [AtD Developers Google Group](http://groups.google.com/group/atd-developers) for discussion and community support.  

When you run your own server, replace `service.afterthedeadline.com` with your server's hostname.

### Privacy:

Your text is sent to a remote server for processing. The system uses too much memory for you to
run it locally. The default server is run by Automattic (http://www.automattic.com) and 
AtD/OpenOffice.org is configured to communicate to this default server over SSL. Automattic does
not keep your text. Our server processes your text and report errors to you. That's it.

If you're not comfortable with this, you may download the open source proofreading server and run
it yourself. http://open.afterthedeadline.com. Configure this extension to use your server under: 
Tools -> Extension Manager -> After the Deadline -> Options

### Contribute

We (Automattic) are no longer supporting this extension.  This code has always been open source.  We're putting it on Github so that you can feel free to fork it, hack it, and release your own version.

Join the [atd-developers](http://groups.google.com/group/atd-developers) list for support.

### License

See `licenses/*` for licenses on individual pieces. The bulk of this extension is derived from the
LanguageTool Open Office extension (http://www.languagetool.org). This extension is released under
the GNU LGPL.

### Author

Raphael Mudge


CHANGELOG.md
### 0.4 - 11 Jan 12

* AtD/OpenOffice now respects the rate limit on the free hosted server
  at service.afterthedeadline.com by spacing proofreading requests out
  to no more than 1 per second.  Proofreading still works as the current 
  line is sent repeatedly to the server and errors will be flagged on
  the next request.  Also removes the error dialog from the rate limiter 
  when using the free hosted server (response code 302).

### 0.3 - 9 Aug 10

* After the Deadline now checks documents set to English (UK). My apologies for
  this. I had the locale wrong (en_UK instead of en_GB). Thanks to Chalkie for
  providing the necessary clue. I've added every English related locale I could 
  find to the code. Hopefully this issue is put to rest now.
* AtD now trims the trailing slash out of the AtD service URL option, if it's
  present. Trailing slashes in the service URL would cause the request to fail
* AtD now resets the AtD service URL to the default if the value is invalid. 
* Added instructions on how to reset the AtD service URL to the AtD options
  dialog (simply clear the URL and hit Save, this will reset it).

### 0.2 - 15 Jul 10

* AtD/OpenOffice.org will now show a dialog with an error when it can't connect
  to the AtD service.
* Added more en locales to the getLocales() call in Main. This may allow AtD to
  check documents with their language set to en_UK 

### 0.1 - 10 Jun 10

* Initial (beta) release.


test.sh
#
# test the AtD client
#

java -classpath lib/\*:atd-openoffice.jar com.afterthedeadline.client.Client


src/com/afterthedeadline/client/Client.java
package com.afterthedeadline.client;

import sleep.runtime.*;
import sleep.interfaces.*;
import sleep.bridges.*;
import sleep.error.*;

import com.afterthedeadline.openoffice.ProofreadingError;

import java.util.*;
import java.io.*;

/* I'm questioning the sanity of this too, but it's a lot quicker to prototype something
   like this in Sleep vs. Java and I'd rather get this extension done. */
public class Client implements RuntimeWarningWatcher {

	public void processScriptWarning(ScriptWarning warning) { 
		System.err.println(warning);
	}

	public static void main(String args[]) {
		System.out.println("Initializing client");
		
		Client temp = new Client();
		ProofreadingError[] errors = temp.proofreadText("This iss an test, the ball was thrown.");
		for (int x = 0; x < errors.length; x++) {
			System.err.println(errors[x]);
		}

		System.out.println("You should have seen two errors");
	}

	public ProofreadingError[] proofreadText(String paragraph) {
		if (!isReady())
			return new ProofreadingError[0];

		Stack args = new Stack();
		args.push(SleepUtils.getScalar(paragraph));

		Scalar temp = script.callFunction("&proofread", args);

		if (!SleepUtils.isEmptyScalar(temp)) {
			ProofreadingError[] errors = new ProofreadingError[temp.getArray().size()];
			Iterator i = temp.getArray().scalarIterator();
			for (int x = 0; i.hasNext(); x++) {
				errors[x] = (ProofreadingError) ((Scalar)i.next()).objectValue();
			}

			return errors;
		}
		else {
			return new ProofreadingError[0];
		}
	}

	protected ScriptInstance script = null;

	public boolean isReady() {
		return script != null;
	}

	public Client() {
		ScriptLoader loader = new ScriptLoader(); 
		Hashtable environment = new Hashtable(); 
		ScriptVariables variables = new ScriptVariables(); 
	
		try { 
			ScriptInstance a = loader.loadScript("webservice.sl", this.getClass().getResourceAsStream("/scripts/webservice.sl"), environment);
			a.setScriptVariables(variables); 
			a.addWarningWatcher(this);


			ScriptInstance b = loader.loadScript("atdclient.sl", this.getClass().getResourceAsStream("/scripts/atdclient.sl"), environment); 
			b.setScriptVariables(variables); 
			b.addWarningWatcher(this);

			a.runScript();
			b.runScript();

			script = b;
		} 
		catch (YourCodeSucksException syntaxErrors) { 
			System.err.println(syntaxErrors.formatErrors()); 
		} 
		catch (Exception ex) {
			System.err.println(ex);
			ex.printStackTrace();
		}
	}	
}


src/com/afterthedeadline/client/Configuration.java
package com.afterthedeadline.client;

import java.util.*;
import java.io.*;
import java.lang.ref.*;

/* loads AtD config file */
public class Configuration {
	protected Properties config;
	protected File       file;
	protected Set        categories;
	protected Set        phrases;
	protected List       listeners = new LinkedList();

	protected static Configuration singleton = null;

	public void fireChange() {
		Iterator i = listeners.iterator();
		while (i.hasNext()) {
			WeakReference ref = (WeakReference)i.next();
			Object o = ref.get();
			if (o == null)
				i.remove();
			else
				((ChangeListener)o).settingsChanged();
		}
	}

	public void addChangeListener(ChangeListener l) {
		listeners.add(new WeakReference(l));
	}

	public synchronized static Configuration getConfiguration() {
		if (singleton == null) {
			singleton = new Configuration(new File(System.getProperty("user.home"), ".AtD-OpenOffice.org"));
			singleton.load();
		}
		return singleton;
	}

	public Configuration(File _file) {
		config = new Properties();
		file   = _file;
	}

	public void load() {
		try {
			config.load(new FileInputStream(file));
			phrases = getIgnoredPhrases();
			categories = getCategories();
		}
		catch (Exception ex) {
			phrases = new HashSet();
			categories = new HashSet();
		}
	}

	private Set<String> createSet(String[] strings) {
		Set temp = new HashSet<String>();
		for (int x = 0; x < strings.length; x++) {
			temp.add(strings[x]);
		}
		return temp;
	}

	public synchronized boolean isIgnored(String phrase) {
		return phrases.contains(phrase);
	}

	public synchronized boolean isEnabled(String category) {
		return categories.contains(category);
	}

	public synchronized void ignorePhrase(String phrase) {
		phrases.add(phrase);
		config.setProperty("ignoredPhrases", createString(phrases));
	}

	public synchronized void removePhrase(String phrase) {
		phrases.remove(phrase);
		config.setProperty("ignoredPhrases", createString(phrases));
	}

	public synchronized void showCategory(String category) {
		categories.add(category);
		config.setProperty("categories", createString(categories));
	}

	public synchronized void hideCategory(String category) {
		categories.remove(category);
		config.setProperty("categories", createString(categories));
	}

	private String createString(Set<String> strings) {
		StringBuffer temp = new StringBuffer();
		Iterator<String> i = strings.iterator();
		while (i.hasNext()) {
			String value = i.next();
			temp.append(value);

			if (i.hasNext())
				temp.append(", ");
		}

		return temp.toString();
	}

	public synchronized Set<String> getIgnoredPhrases() {
		return createSet(config.getProperty("ignoredPhrases", "").split(",\\s+"));
	}

	public synchronized Set<String> getCategories() {
		return createSet(config.getProperty("categories", "").split(",\\s+"));
	}

	public synchronized String getServiceHost() {
		if ("".equals(config.getProperty("host")) || config.getProperty("host") == null) {
			return "https://en.service.afterthedeadline.com";
		}

		String host = (config.getProperty("host") + "").trim();
		if (host.endsWith("/")) {
			host = host.substring(0, host.length() - 1);
		}

		return host;
	}

	public synchronized void setServiceHost(String name) {
		config.setProperty("host", name);
	}

	public void save() {
		try {
			config.save(new FileOutputStream(file), "AtD-OpenOffice Properties");
			fireChange();
		}
		catch (Exception ex) {
			throw new RuntimeException("Could not save properties\nLocation:" + file + "\n" + ex.getMessage());
		}
	}
}


src/com/afterthedeadline/openoffice/SingletonFactory.java
/* After the Deadline, a proofreading tool
 * Copyright (C) 2009-2010 Automattic (http://www.automattic.com)
 *
 * AtDdev: Raphael Mudge
 *
 * This code is derived from:
 *
 * LanguageTool, a natural language style checker 
 * Copyright (C) 2006 Daniel Naber (http://www.danielnaber.de)
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
 * USA
 */
package com.afterthedeadline.openoffice;

import com.sun.star.lang.XSingleComponentFactory;
import com.sun.star.uno.XComponentContext;

/**
 * This class is a factory that creates only a single instance,
 * or a singleton, of the Main class. Used for performance 
 * reasons and to allow various parts of code to interact.
 *
 * @author Marcin Miłkowski
 */
public class SingletonFactory implements XSingleComponentFactory {

  private transient com.afterthedeadline.openoffice.Main instance;

  public final Object createInstanceWithArgumentsAndContext(final Object[] arguments, 
      final XComponentContext xContext) throws com.sun.star.uno.Exception {    
    return createInstanceWithContext(xContext);
  }

  public final Object createInstanceWithContext(final XComponentContext xContext) throws com.sun.star.uno.Exception {    
    if (instance == null) {     
      instance = new com.afterthedeadline.openoffice.Main(xContext);      
    } else {  
      instance.changeContext(xContext);      
    }
    return instance;
  }  
}


src/com/afterthedeadline/openoffice/CentralRegistrationClass.java
/*
 * CentralRegistrationClass.java
 *
 * This class was automatically created by the OpenOffice.org Wizard.
 */

package com.afterthedeadline.openoffice;

import com.sun.star.lang.XSingleComponentFactory;
import com.sun.star.registry.XRegistryKey;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.JarURLConnection;
import java.net.URL;
import java.util.Enumeration;
import java.util.StringTokenizer;
import java.util.jar.Attributes;
import java.util.jar.Manifest;

/**
 *
 * @author chris
 */
public class CentralRegistrationClass {

    public static XSingleComponentFactory __getComponentFactory( String sImplementationName ) {
        String regClassesList = getRegistrationClasses();
        StringTokenizer t = new StringTokenizer(regClassesList, " ");
        while (t.hasMoreTokens()) {
            String className = t.nextToken();
            if (className != null && className.length() != 0) {
                try {
                    Class regClass = Class.forName(className);
                    Method writeRegInfo = regClass.getDeclaredMethod("__getComponentFactory", new Class[]{String.class});
                    Object result = writeRegInfo.invoke(regClass, sImplementationName);
                    if (result != null) {
                       return (XSingleComponentFactory)result;
                    }
                }
                catch (ClassNotFoundException ex) {
                    ex.printStackTrace();
                } catch (ClassCastException ex) {
                    ex.printStackTrace();
                } catch (SecurityException ex) {
                    ex.printStackTrace();
                } catch (NoSuchMethodException ex) {
                    ex.printStackTrace();
                } catch (IllegalArgumentException ex) {
                    ex.printStackTrace();
                } catch (InvocationTargetException ex) {
                    ex.printStackTrace();
                } catch (IllegalAccessException ex) {
                    ex.printStackTrace();
                }
            }
        }
        return null;
    }

    public static boolean __writeRegistryServiceInfo( XRegistryKey xRegistryKey ) {
        boolean bResult = true;
        String regClassesList = getRegistrationClasses();
        StringTokenizer t = new StringTokenizer(regClassesList, " ");
        while (t.hasMoreTokens()) {
            String className = t.nextToken();
            if (className != null && className.length() != 0) {
                try {
                    Class regClass = Class.forName(className);
                    Method writeRegInfo = regClass.getDeclaredMethod("__writeRegistryServiceInfo", new Class[]{XRegistryKey.class});
                    Object result = writeRegInfo.invoke(regClass, xRegistryKey);
                    bResult &= ((Boolean)result).booleanValue();
                }
                catch (ClassNotFoundException ex) {
                    ex.printStackTrace();
                } catch (ClassCastException ex) {
                    ex.printStackTrace();
                } catch (SecurityException ex) {
                    ex.printStackTrace();
                } catch (NoSuchMethodException ex) {
                    ex.printStackTrace();
                } catch (IllegalArgumentException ex) {
                    ex.printStackTrace();
                } catch (InvocationTargetException ex) {
                    ex.printStackTrace();
                } catch (IllegalAccessException ex) {
                    ex.printStackTrace();
                }
            }
        }
        return bResult;
    }

    private static String getRegistrationClasses() {
	/* this is a hack because the method wasn't finding the appropriate entry in the manifest */
        return "com.afterthedeadline.openoffice.Main com.afterthedeadline.openoffice.config.DialogEventHandler";
    }

    /** Creates a new instance of CentralRegistrationClass */
    private CentralRegistrationClass() {
    }
}


src/com/afterthedeadline/openoffice/Main.java
/* After the Deadline, a proofreading tool
 * Copyright (C) 2009-2010 Automattic (http://www.automattic.com)
 *
 * AtDdev: Raphael Mudge
 *
 * This code is derived from:
 *
 * LanguageTool, a natural language style checker 
 * Copyright (C) 2005 Daniel Naber (http://www.danielnaber.de)
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
 * USA
 */

package com.afterthedeadline.openoffice;

import java.io.File;
import java.util.*;

import javax.swing.JOptionPane;
import javax.swing.UIManager;

import com.afterthedeadline.client.*;

import com.sun.star.awt.XWindow;
import com.sun.star.awt.XWindowPeer;
import com.sun.star.beans.PropertyValue;
import com.sun.star.beans.XPropertySet;
import com.sun.star.frame.XDesktop;
import com.sun.star.frame.XModel;
import com.sun.star.lang.IllegalArgumentException;
import com.sun.star.lang.Locale;
import com.sun.star.lang.XComponent;
import com.sun.star.lang.XMultiComponentFactory;
import com.sun.star.lang.XServiceDisplayName;
import com.sun.star.lang.XServiceInfo;
import com.sun.star.lang.XSingleComponentFactory;
import com.sun.star.lib.uno.helper.Factory;
import com.sun.star.lib.uno.helper.WeakBase;
import com.sun.star.linguistic2.ProofreadingResult;
import com.sun.star.linguistic2.SingleProofreadingError;
import com.sun.star.linguistic2.XLinguServiceEventBroadcaster;
import com.sun.star.linguistic2.XLinguServiceEventListener;
import com.sun.star.linguistic2.XProofreader;
import com.sun.star.registry.XRegistryKey;
import com.sun.star.task.XJobExecutor;
import com.sun.star.text.XTextViewCursor;
import com.sun.star.text.XTextViewCursorSupplier;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XComponentContext;

public class Main extends WeakBase implements XJobExecutor, XServiceDisplayName, XServiceInfo, XProofreader, XLinguServiceEventBroadcaster, ChangeListener {
	private String docID;

	private Set<String> disabledRules = new HashSet();
	private List<XLinguServiceEventListener> xEventListeners;

	private int position;

	/**
	 * Service name required by the OOo API && our own name.
	 */
	private static final String[] SERVICE_NAMES = { "com.sun.star.linguistic2.Proofreader", "com.afterthedeadline.openoffice.Main" };

	private XComponentContext xContext;
	private Client atdClient = null;

	public Main(final XComponentContext xCompContext) {
		try {
			changeContext(xCompContext);
			disabledRules = new HashSet<String>();
			xEventListeners = new ArrayList<XLinguServiceEventListener>();
			atdClient = new Client();
			Configuration.getConfiguration().addChangeListener(this);
		} 
		catch (final Throwable t) {
			showError(t);
		}
	}

	public final void changeContext(final XComponentContext xCompContext) {
		xContext = xCompContext;
	}

	private XComponent getxComponent() {
		try {
			final XMultiComponentFactory xMCF = xContext.getServiceManager();
			final Object desktop = xMCF.createInstanceWithContext("com.sun.star.frame.Desktop", xContext);
			final XDesktop xDesktop = (XDesktop) UnoRuntime.queryInterface(XDesktop.class, desktop);
			final XComponent xComponent = xDesktop.getCurrentComponent();
			return xComponent;
		} 
		catch (final Throwable t) {
			showError(t);
			return null;
		}
	}

	/**
	 * Runs the grammar checker on paragraph text.
	 * 
	 * @param String
	 *          docID - document ID
	 * @param String
	 *          paraText - paragraph text
	 * @param locale
	 *          Locale - the text Locale
	 * @param startOfSentencePos start of sentence position
	 * @param nSuggestedBehindEndOfSentencePosition end of sentence position
	 * @param PropertyValue
	 *          [] props - properties
	 * @return ProofreadingResult containing the results of the check. *
	 * @throws IllegalArgumentException
	 *           (not really, LT simply returns the ProofreadingResult with the
	 *           values supplied)
	 */
	public final ProofreadingResult doProofreading(final String docID, final String paraText, final Locale locale, final int startOfSentencePos, final int nSuggestedBehindEndOfSentencePosition, final PropertyValue[] props) {
		final ProofreadingResult paRes = new ProofreadingResult();
		try {
			paRes.nStartOfSentencePosition = startOfSentencePos;
			paRes.xProofreader = this;
			paRes.aLocale = locale;
			paRes.aDocumentIdentifier = docID;
			paRes.aText = paraText;
			paRes.aProperties = props;
			return doGrammarCheckingInternal(paraText, locale, paRes);
		} 
		catch (final Throwable t) {
			showError(t);
			return paRes;
		}
	}

	synchronized private final ProofreadingResult doGrammarCheckingInternal(final String paraText, final Locale locale, final ProofreadingResult paRes) {
		// process enabled rules, disabled rules, and categories here
        
		try {
			String sentence = paraText; 
			paRes.nStartOfSentencePosition = position;
			paRes.nStartOfNextSentencePosition = position + sentence.length();
			paRes.nBehindEndOfSentencePosition = paRes.nStartOfNextSentencePosition;

			ProofreadingError[] errors = atdClient.proofreadText(paraText);
			
			List errorsList = new LinkedList();
			for (int x = 0; x < errors.length; x++) {
				SingleProofreadingError aError = createOOoError(errors[x]);
				if (!disabledRules.contains(aError.aRuleIdentifier))
					errorsList.add(aError);
			}

			SingleProofreadingError[] errorArray = new SingleProofreadingError[errorsList.size()];
			Iterator i = errorsList.iterator();
			for (int x = 0; i.hasNext(); x++) {
				errorArray[x] = (SingleProofreadingError)i.next();
			}

			paRes.aErrors = errorArray;
		} 
		catch (final Throwable t) {
			showError(t);
			paRes.nBehindEndOfSentencePosition = paraText.length();
		}      
		return paRes;
	}

	/**
	 * Creates a SingleGrammarError object for use in OOo.
	 * @param myMatch
	 *          ruleMatch - LT rule match
	 * 
	 * @return SingleGrammarError - object for OOo checker integration
	 */
	private SingleProofreadingError createOOoError(ProofreadingError error) {
		final SingleProofreadingError aError = new SingleProofreadingError();
		
		if ("spelling".equals(error.type) || "Hyphen Required".equals(error.description)) {
			/* OOo ignores this value setting it to PROOFREADING, thanks OOo, not really */
                        aError.nErrorType = com.sun.star.text.TextMarkupType.SPELLCHECK;
		}
		else {
			aError.nErrorType = com.sun.star.text.TextMarkupType.PROOFREADING;
		}

		// the API currently has no support for formatting text in comments

		aError.aFullComment = error.description;
		aError.aShortComment = error.description;

		if (error.suggestions != null) {
			aError.aSuggestions = new String[error.suggestions.size()];

			Iterator i = error.suggestions.iterator();
			for (int y = 0; i.hasNext(); y++) {
				aError.aSuggestions[y] = i.next().toString();
			}
		}
		else {
			aError.aSuggestions = new String[0];
		}

		aError.nErrorStart = error.locationStart;
		aError.nErrorLength = error.errorString.length();
		aError.aRuleIdentifier = error.description;
		aError.aProperties = new PropertyValue[0];

		return aError;
	}

	public boolean hasLocale(Locale locale) {
		return "en".equals( locale.Language );
	}

	public final Locale[] getLocales() {
		return new Locale[] { 
			new Locale("en", "us", "en_US"), 
			new Locale("en", "ca", "en_CA"), 
			new Locale("en", "uk", "en_UK"), 
			new Locale("en", "gb", "en_GB"), 
			new Locale("en", "in", "en_IN"), 
			new Locale("en", "au", "en_AU"), 
			new Locale("en", "nz", "en_NZ"), 
			new Locale("en", "za", "en_ZA"), 
			new Locale("en", "zw", "en_ZW"), 
			new Locale("en", "ph", "en_PH"), 
			new Locale("en", "be", "en_BE"), 
			new Locale("en", "bw", "en_BW"), 
			new Locale("en", "bz", "en_BZ"), 
			new Locale("en", "jm", "en_JM"), 
			new Locale("en", "be", "en_BE"), 
			new Locale("en", "gu", "en_GU"), 
			new Locale("en", "hk", "en_HK"), 
			new Locale("en", "mh", "en_MH"), 
			new Locale("en", "mp", "en_MP"), 
			new Locale("en", "pk", "en_PK"), 
			new Locale("en", "sg", "en_SG"), 
			new Locale("en", "tt", "en_TT"), 
			new Locale("en", "um", "en_UM"), 
			new Locale("en", "vi", "en_VI")
		};
	}

	public final boolean isSpellChecker() {
		return false;
	}

	public final boolean hasOptionsDialog() {
		return false;
	}

	/**
	 * Add a listener that allow re-checking the document after changing the
	 * options in the configuration dialog box.
	 * 
	 * @param xLinEvLis
	 *          - the listener to be added
	 * @return true if listener is non-null and has been added, false otherwise.
	 */
	public final boolean addLinguServiceEventListener(final XLinguServiceEventListener xLinEvLis) {
		if (xLinEvLis == null) {
			return false;
		}
		xEventListeners.add(xLinEvLis);
		return true;
	}

	/**
	 * Remove a listener from the event listeners list.
	 * 
	 * @param xLinEvLis
	 *          - the listener to be removed
	 * @return true if listener is non-null and has been removed, false otherwise.
	 */
	public final boolean removeLinguServiceEventListener(final XLinguServiceEventListener xLinEvLis) {
		if (xLinEvLis == null) {
			return false;
		}
    
		if (xEventListeners.contains(xLinEvLis)) {
			xEventListeners.remove(xLinEvLis);
			return true;
		}
		return false;
	}

	public final void recheckDocument() {
		if (!xEventListeners.isEmpty()) {
			for (final XLinguServiceEventListener xEvLis : xEventListeners) {
				if (xEvLis != null) {
					final com.sun.star.linguistic2.LinguServiceEvent xEvent = new com.sun.star.linguistic2.LinguServiceEvent();
					xEvent.nEvent = com.sun.star.linguistic2.LinguServiceEventFlags.PROOFREAD_AGAIN;
					xEvLis.processLinguServiceEvent(xEvent);
 				}
			}
		}
	}

	public final void resetDocument() {
		disabledRules = new HashSet<String>();
		recheckDocument();
	}

	public String[] getSupportedServiceNames() {
		return getServiceNames();
	}

	public static String[] getServiceNames() {
		return SERVICE_NAMES;
	}

	public boolean supportsService(final String sServiceName) {
		for (final String sName : SERVICE_NAMES) {
			if (sServiceName.equals(sName)) {
				return true;
			}
		}
		return false;
	}

	public String getImplementationName() {
		return Main.class.getName();
	}

	public static XSingleComponentFactory __getComponentFactory(final String sImplName) {
		SingletonFactory xFactory = null;
		if (sImplName.equals(Main.class.getName())) {
			xFactory = new SingletonFactory();
		}
		return xFactory;
	}

	public static boolean __writeRegistryServiceInfo(final XRegistryKey regKey) {
		return Factory.writeRegistryServiceInfo(Main.class.getName(), Main.getServiceNames(), regKey);
	}

	public void trigger(final String sEvent) {
		if (!javaVersionOkay()) {
			return;
		}

		try {
			if (sEvent.equals("reset")) {
				resetDocument();	
			}
			else {
				System.err.println("Sorry, don't know what to do, sEvent = " + sEvent);
			}
		} 
		catch (final Throwable e) {
			showError(e);
		}
	}

	public void settingsChanged() {
		resetDocument();
	}

	private boolean javaVersionOkay() {
		final String version = System.getProperty("java.version");

		if (version != null && (version.startsWith("1.0") || version.startsWith("1.1") || version.startsWith("1.2") || version.startsWith("1.3") || version.startsWith("1.4"))) {
			final DialogThread dt = new DialogThread("Error: LanguageTool requires Java 1.5 or later. Current version: " + version);
			dt.start();
			return false;
		}

		return true;
	}

	static void showError(final Throwable e) {
		String metaInfo = "OS: " + System.getProperty("os.name") + " on " + System.getProperty("os.arch") + ", Java version " + System.getProperty("java.vm.version") + " from " + System.getProperty("java.vm.vendor");
		String msg = "An error has occurred in After the Deadline:\n" + e.toString() + "\nStacktrace:\n";
    
		final StackTraceElement[] elem = e.getStackTrace();
		for (final StackTraceElement element : elem) {
			msg += element.toString() + "\n";
		}
		msg += metaInfo;
		final DialogThread dt = new DialogThread(msg);
		dt.start();
	}

	public static void showMessage(String msg) {
		final DialogThread dt = new DialogThread(msg);
		dt.start();
	}

	public void ignoreRule(final String ruleId, final Locale locale) throws IllegalArgumentException {
		try {
			disabledRules.add(ruleId);
			recheckDocument();
		} 
		catch (final Throwable t) {
			showError(t);
		}
  	}

	public void resetIgnoreRules() {
		try {
			disabledRules = new HashSet<String>();
		} 
		catch (final Throwable t) {
			showError(t);
		}
	}
  
	public String getServiceDisplayName(Locale locale) {
		return "After the Deadline";
	}
}

class DialogThread extends Thread {
	final private String text;

	DialogThread(final String text) {
		this.text = text;
	}

	@Override
	public void run() {
		JOptionPane.showMessageDialog(null, text);
	}
}
